pandas/_libs/parsing.pyx

```diff
@@ -0,0 +1,682 @@
+_MONTH_ALIASES = {(k + 1): v for k, v in enumerate(_MONTHS)}
+
+cdef object _TIMEPAT = re.compile(r'^([01]?[0-9]|2[0-3]):([0-5][0-9])')
+
+cdef set _not_datelike_strings = set(['a', 'A', 'm', 'M', 'p', 'P', 't', 'T'])
+
+NAT_SENTINEL = object()
+# This allows us to reference NaT without having to import it
+
+
+def parse_datetime_string(date_string, freq=None, dayfirst=False,
+                          yearfirst=False, **kwargs):
+    """parse datetime string, only returns datetime.
+    Also cares special handling matching time patterns.
+
+    Returns
+    -------
+    datetime
+    """
+
+    cdef:
+        object dt
+
+    if not _does_string_look_like_datetime(date_string):
+        raise ValueError('Given date string not likely a datetime.')
+
+    if _TIMEPAT.match(date_string):
+        # use current datetime as default, not pass _DEFAULT_DATETIME
+        dt = du_parse(date_string, dayfirst=dayfirst,
+                      yearfirst=yearfirst, **kwargs)
+        return dt
+
+    try:
+        dt, _, _ = _parse_dateabbr_string(date_string, _DEFAULT_DATETIME, freq)
+        return dt
+    except DateParseError:
+        raise
+    except ValueError:
+        pass
+
+    try:
+        dt = du_parse(date_string, default=_DEFAULT_DATETIME,
+                      dayfirst=dayfirst, yearfirst=yearfirst, **kwargs)
+    except TypeError:
+        # following may be raised from dateutil
+        # TypeError: 'NoneType' object is not iterable
+        raise ValueError('Given date string not likely a datetime.')
+
+    return dt
+
+
+def parse_time_string(arg, freq=None, dayfirst=None, yearfirst=None):
+    """
+    Try hard to parse datetime string, leveraging dateutil plus some extra
+    goodies like quarter recognition.
+
+    Parameters
+    ----------
+    arg : compat.string_types
+    freq : str or DateOffset, default None
+        Helps with interpreting time string if supplied
+    dayfirst : bool, default None
+        If None uses default from print_config
+    yearfirst : bool, default None
+        If None uses default from print_config
+
+    Returns
+    -------
+    datetime, datetime/dateutil.parser._result, str
+    """
+    if not isinstance(arg, string_types):
+        return arg
+
+    if getattr(freq, "_typ", None) == "dateoffset":
+        freq = freq.rule_code
+
+    if dayfirst is None:
+        from pandas.core.config import get_option
+        dayfirst = get_option("display.date_dayfirst")
+    if yearfirst is None:
+        from pandas.core.config import get_option
+        yearfirst = get_option("display.date_yearfirst")
+
+    res = parse_datetime_string_with_reso(arg, freq=freq,
+                                          dayfirst=dayfirst,
+                                          yearfirst=yearfirst)
+    if res[0] is NAT_SENTINEL:
+        from pandas._libs.tslib import NaT
+        res = (NaT,) + res[1:]
+    return res
+
+
+def parse_datetime_string_with_reso(date_string, freq=None, dayfirst=False,
```

Comment:
maybe not for this PR if we want to keep it 'moving only', but after moving it would make sense to me to combine this with the function above, as this is the only place where it is used